interfaces/apparmor/spec.go

```diff
@@ -146,6 +166,70 @@ func (spec *Specification) AddSnippet(snippet string) {
 	}
 }
 
+// RegisterSnippetKey adds a key to the list of valid keys
+func (spec *Specification) RegisterSnippetKey(key interfaces.SnippetKey) {
+	if spec.snippetKeys == nil {
+		spec.snippetKeys = make(map[interfaces.SnippetKey]bool)
+	}
+	if _, ok := spec.snippetKeys[key]; ok {
+		logger.Panicf("priority key %s is already registered", key.String())
+	}
+	spec.snippetKeys[key] = true
+}
+
+// AddPrioritizedSnippet adds a new apparmor snippet to all applications and hooks using the interface,
+// but identified with a key and a priority. If no other snippet exists with that key, the snippet is
+// added like with AddSnippet, but if there is already another snippet with that key, the priority of
+// both will be taken into account to decide whether the new snippet replaces the old one, is appended
+// to it, or is just ignored. The key must have been previously registered using RegisterSnippetKey().
+func (spec *Specification) AddPrioritizedSnippet(snippet string, key interfaces.SnippetKey, priority uint) {
+	if _, ok := spec.snippetKeys[key]; !ok {
+		logger.Panicf("priority key %s is not registered", key.String())
+	}
+	if len(spec.securityTags) == 0 {
+		return
+	}
+	if spec.prioritizedSnippets == nil {
+		spec.prioritizedSnippets = make(map[string]map[interfaces.SnippetKey]prioritizedSnippets)
+	}
+
+	for _, tag := range spec.securityTags {
+		if _, exists := spec.prioritizedSnippets[tag]; !exists {
+			spec.prioritizedSnippets[tag] = make(map[interfaces.SnippetKey]prioritizedSnippets)
+		}
+		if snippets, exists := spec.prioritizedSnippets[tag][key]; exists {
```

Comment:
This is not a biggie but perhaps it could be more readable this way. Note that priority is uint so zero value is also the smallest value, which is nice.

```go
snippets, ok := spec.prioritizedSnippets[tag][key]
// Zero value has zero priority
if snippets.priority == priority {
   snippets.snippets = append(snippets.snippets, snippet)
} else if snippets.priority < priority {
   snippets.snippets = append([]string(nil), snippet)
} // smaller priority, discard
spec.prioritizedSnippets[tag][key] = snippets
```

Separately if you rename `prioritizedSnippets.snippets` to `prioritizedSnippets.list` then the code might read better in this expression.

Reply:
I see what you mean... I like the concept, but I'll add some comments to make it more clear.